### PR TITLE
[US1537] snapshot related tasks

### PR DESCRIPTION
**Pull request by vishnuitta**
This PR addresses the below snapshot related tasks:

    implement 'snapcreate' and 'snapdestroy' options for istgtcontrol
    API to pause write IOs, waits for any ongoing write IOs to finish
    API to send snapshot create/destroy commands to replica
    API to create and destroy snapshots by using all above functions

Testcases for this feature

    Verifying if volume exists for correct volname
    Verifying if volume does not exist for wrong volname
    Verifying status of volume as degraded with less number of replicas connected
    Verifying volume state is healthy with required number of replicas connected and being in healthy state
    Hitting the case of timeout while waiting for ongoing write IOs being served, and verify that write IOs are resumed and error is returned from API
    Verifying that configured amount of time has not been passed by the time snapshot IO is being sent
    Hit the case of adding snapshot command to the list, and verify command is sent eventually
    Verify that snapshots are destroyed in case of one or multiple failures from replica
    Verify that write IOs are resumed in case of one or multiple failures from replica
    Hit the case of timeout while waiting for snapshot response(s) and verify that write IOs are resumed, snapshots are destroyed and error is returned
    Verify success case of creating snapshot, write IOs are resumed and returned 0 as success from API

Created "TS2" in rally to track more testcases.

### CODE REVIEW

File: src/replication.c [hunk 1]
```diff
@@ -448,6 +449,245 @@ send_replica_handshake_query(replica_t *replica, spec_t *spec)
...
+	replica_t *replica;
+	int free_rcomm_mgmt = 0;
+	struct timespec last, now, diff;
+	struct rcommon_mgmt_cmd *rcomm_mgmt;
+	int rc;
+
+	clock_gettime(CLOCK_MONOTONIC, &last);
+	MTX_LOCK(&spec->rq_mtx);
+
+	if (is_volume_healthy(spec) == false) {
```
Comment (satbirchhikara) [inline src/replication.c]: You can move volume healthy check in func pause_and_timed_wait_for_ongoing_ios and so does MTX_LOCK. At line 587, we are unlocking it without locking it in the same scope which is little difficult to maintain going forward.

File: src/istgt_integration.h [hunk 2]
```diff
@@ -60,6 +60,7 @@ typedef struct replica_s {
 	int data_eventfd;
 	int epfd;
 	int dont_free;
+	uint64_t inflight_write_io_cnt;
```
[the remark this answers is not in the document]
Comment (vishnuitta) [reply src/istgt_integration.h]: done

File: src/data_conn.c [hunk 3]
```diff
@@ -61,6 +61,8 @@ static rcmd_t *dequeue_replica_cmdq(replica_t *replica);
 		rcomm_cmd = rcmd->rcommq_ptr; 	\
 		_mtx = rcomm_cmd->mutex;	\
 		_cond = rcomm_cmd->cond_var;	\
+		if (rcomm_cmd->opcode == ZVOL_OPCODE_WRITE)	\
+			r->inflight_write_io_cnt -= 1;	\
```
[the remark this answers is not in the document]
Comment (vishnuitta) [reply src/data_conn.c]: done

File: src/istgt_lu_disk.c [hunk 4]
```diff
@@ -5917,6 +5917,11 @@ istgt_lu_disk_lbwrite(ISTGT_LU_DISK *spec, CONN_Ptr conn, ISTGT_LU_CMD_Ptr lu_cm
 		iov[i].iov_len = lu_cmd->iobuf[i].iov_len;
 	}
 
+	while (spec->lu->quiesce) {
+		ISTGT_ERRLOG("c#%d LU%d: quiescing write IOs\n", conn->id, spec->lu->num);
```
[the remark this answers is not in the document]
Comment (vishnuitta) [reply src/istgt_lu_disk.c]: this quiesces write IOs on volume as per the print says.

File: src/istgt_lu_disk.c [hunk 5]
```diff
@@ -5963,6 +5968,10 @@ istgt_lu_disk_lbwrite(ISTGT_LU_DISK *spec, CONN_Ptr conn, ISTGT_LU_CMD_Ptr lu_cm
 			sleep(8);
 		}
 
+	MTX_LOCK(&spec->rq_mtx);
+	spec->inflight_write_io_cnt += 1;
```
[the remark this answers is not in the document]
Comment (vishnuitta) [reply src/istgt_lu_disk.c]: done. using replica_write_io_count for replica variable, and leaving as it is for spec variable.

File: src/replication.c [hunk 6]
```diff
@@ -805,6 +1044,19 @@ read_io_resp(spec_t *spec, replica_t *replica, io_event_t *revent)
 					free(*resp_data);
 					break;
 
+				case ZVOL_OPCODE_SNAP_CREATE:
+					if ((*resp_data) != NULL)
```
[the remark this answers is not in the document]
Comment (vishnuitta) [reply src/replication.c]: debug assert builds are not available for istgt. we need to create one with asserts wherever required.
In this case, removing the check.

File: src/replication.c [hunk 7]
```diff
@@ -805,6 +1044,19 @@ read_io_resp(spec_t *spec, replica_t *replica, io_event_t *revent)
...
+						REPLICA_ERRLOG("resp_data should be NULL for snap create\n");
+
+					/* snap create response must come from mgmt connection */
+					handle_snap_create_resp(replica, mgmt_cmd);
+					break;
+
+				case ZVOL_OPCODE_SNAP_DESTROY:
+					if ((*resp_data) != NULL)
```
[the remark this answers is not in the document]
Comment (vishnuitta) [reply src/replication.c]: done. removed.

File: src/replication.c [hunk 8]
```diff
@@ -448,6 +449,245 @@ send_replica_handshake_query(replica_t *replica, spec_t *spec)
...
+	MTX_LOCK(&rcomm_mgmt->mtx);
+	if (hdr->status != ZVOL_OP_STATUS_OK)
+		rcomm_mgmt->cmds_failed++;
+	else
+		rcomm_mgmt->cmds_succeeded++;
+	if ((rcomm_mgmt->caller_gone == 1) &&
+	    (rcomm_mgmt->cmds_sent == (rcomm_mgmt->cmds_failed + rcomm_mgmt->cmds_succeeded)))
+		delete = true;
+	MTX_UNLOCK(&rcomm_mgmt->mtx);
+	if (delete == true)
```
[the remark this answers is not in the document]
Comment (vishnuitta) [reply src/replication.c]: yes.. we need to increase the cmds_failed. A task is available under US1536 to handle this.
Probably, I can cover it in this PR itself. Will include this in next commit on this PR.

File: src/replication.c (hunk 1 above)
Comment (vishnuitta) [reply src/replication.c]: pause_and_timed_.. function does the check for volume health. But, before that, it quiesce IOs also, which is not required if volume is not healthy.
Regarding line 587, comment is available for the function to have lock before calling this function.

**Comment by vishnuitta** [comment]
thanks @satbirchhikara for very good review comments .. incorporated them
let me know if anything more need to be done.

**Pull request merged**